ci(release): Add weekly pre-releases workflow for Generals and GeneralsMD

fbraz3 opened the pull request:
**What this PR does?**

This PR complements https://github.com/TheSuperHackers/GeneralsGameCode/pull/1171

Creates a github action to create a weekly release with the latest code changes. It will be scheduled to every moday.

https://github.com/fbraz3/GeneralsGameCode/actions/runs/15689781515
<img width="1084" alt="image" src="https://github.com/user-attachments/assets/2922624a-2246-4357-9a17-3b6cea422ade" />

It also can be mannualy triggered, and it have parameters to bypass the code changes check to force a new build
<img width="1084" alt="image" src="https://github.com/user-attachments/assets/d01d2d14-80ca-4a9a-8929-f5a9204c9cb4" />

The release notes also have a changelog with all commits since the last release - for the first release it will be limited to the last 10 commits
<img width="1237" alt="image" src="https://github.com/user-attachments/assets/55b6a1da-f3df-4bad-94b1-6911d17b1ad5" />

xezon commented:
What does this change produce then?

roossienb commented:
If these are our 'official' pre-releases then they need to be properly versions.

fbraz3 commented:
> What does this change produce then?

It will produce zip files weekly with the main executables for generals, worldbuilder and w3dview both for both games.

roossienb commented:
> > If these are our 'official' pre-releases then they need to be properly versions.
> 
> I can adjust for a semVer schema if it looks better.

It's not about looking better, it's about having traceability when somebody reports an issue or bug.

fbraz3 commented:
@xezon @roossienb I’ve just updated the pull request to use semantic versioning.

The version is now based on `.github/workflows/base-version.txt` and the patch level is incremented automatically at build time.

When the project is ready for a stable release, we just need to update the version file to `1.0`.

I ran the workflow on my fork and everything went smoothly:
https://github.com/fbraz3/GeneralsGameCode/actions/runs/15241083502
https://github.com/fbraz3/GeneralsGameCode/releases

<img width="1002" alt="image" src="https://github.com/user-attachments/assets/19a82f40-59c2-4095-b127-1af6bff04be3" />
<img width="1178" alt="image" src="https://github.com/user-attachments/assets/f1f1072e-18d8-4a44-ad93-74e7efc1ecda" />

roossienb commented:
In a discussion with @Mauller he told me that they did nightly builds in which they used the latest commit number as the build number in the application. This makes it easy to track back.

fbraz3 commented:
> In a discussion with @Mauller he told me that they did nightly builds in which they used the latest commit number as the build number in the application. This makes it easy to track back.

It’s fine to change the logic to whatever is needed. But just need to decide what will be the pattern.

All approaches have its own pros and cons. Using latest commit it’s a good approach but we can get more than one commit in a day, and this would be a problem to track anyway.

What about to mix the approaches like `0.9.{commit_id}` or `0.9.{release_date}` ?

roossienb commented:
I think most important is that the version is also shown in-game (so when you open the options menu, in the bottom).

fbraz3 commented:
> I think most important is that the version is also shown in-game (so when you open the options menu, in the bottom).

I totally agree with you. But I don’t have enough knowledge in C language. Can you guys help me to do that?

fbraz3 commented:
@xezon @roossienb @OmniBlade @tintinhamans

I've update the first comment of this PR with the latest changes.

xezon commented:
Please move the version change to a new Pull Request because it is not related to pre-releases workflow.

This relates to task #1010

xezon commented:
I don't think we have anyone active right now who can review this CI stuff.

fbraz3 commented:
> I don't think we have anyone active right now who can review this CI stuff.

I work as a devops for many years, maybe I can help the project on CI reviews

fbraz3 commented:
> Please move the version change to a new Pull Request because it is not related to pre-releases workflow.
> 
> This relates to task #1010

Will work on that this weekend.

prettyblueberry commented:
How can I play the pr-release build with GenLauncher to play a mod?

prettyblueberry commented:
> use my fork for now.

I tried already pre-release build of your fork. I replaced generals.exe and modded.exe in working genlauncher directory, but the genlauncher didn't run a mod correctly.

fbraz3 commented:
> > use my fork for now.
> 
> I tried already pre-release build of your fork. I replaced generals.exe and modded.exe in working genlauncher directory, but the genlauncher didn't run a mod correctly. Could you try to run genlauncher using your build for RotR or TEOD mod?

My build uses the exact same code from SuperHacker's repository. Please open an [issue](https://github.com/TheSuperHackers/GeneralsGameCode/issues) for further analysis.

xezon commented:
@OmniBlade can you review this?

xezon commented on .github/workflows/base-version.txt:
```diff
@@ -0,0 +1 @@
+0.9.1
```
Why was this version value chosen?

fbraz3 replied:
It's just a random value and can be set whatever we want to.

xezon reviewed (commented):
Thank you for the images. The resulting Release overview looks pretty good. Perhaps we can name this "Preview Version" instead of "Pre-Release"? Basically we want to communicate that auto published versions are not hand picked final releases and can contain side effects.

xezon commented on .github/workflows/base-version.txt:
```diff
@@ -0,0 +1 @@
+0.9.1
```
I agree. We can make the versioning enterily independent of the CI.

xezon commented:
I do think our previews and releases should only contain the vc6 + tools builds, no win32, no extras. Because the win32 is not meant to be used by players at this time.

fbraz3 commented on .github/workflows/weekly-release.yml:
```diff
@@ -0,0 +1,298 @@
+            echo "changed=true" >> $GITHUB_OUTPUT
+          fi
+
+  calculate-version:
```
This is needed by [build-toolchain](https://github.com/TheSuperHackers/GeneralsGameCode/pull/929/files#diff-19288187e42413b2ca1aa57648719958550f0b6eda5b9a9e3f86d2ec5c105fe6R42) and will be used in C++ counterpart [here](https://github.com/TheSuperHackers/GeneralsGameCode/pull/1171/files#diff-d8a93a9d99fbe11ea093e102711d9d514e2ba1543d290d44fe2243ecfcbd5458R27) and [here](https://github.com/TheSuperHackers/GeneralsGameCode/pull/1171/files#diff-6f3fc17276690acdbe1db84956a3511f549376b6ba785ea90f49902f9ae84ef0R29)

xezon commented:
If the CMake can generate the version, then is it fine to use that instead?

fbraz3 commented:
O can try do 

> If the CMake can generate the version, then is it fine to use that instead?

Although I don’t have experience in c/c++ coding, I can try to do it. 

What will be the version pattern? Just the for version?

xezon commented:
I can do that, because I already have all the code here to facilitate this.

fbraz3 commented:
> I can do that, because I already have all the code here to facilitate this.

Great!

Just to make sure we're on the same page, for this PR I need to:

- Remove the version generation logic from `weekly-release`
- Roll back the version logic in `build-toolchain`
- Change the release title format to `preview-YYYY-mm-dd`
- Remove the Win32 builds

Anything else?

fbraz3 commented:
Hey @xezon sorry for a long delay, i'm facing some personal issues and not had much time last weeks. I Intend to work on CI side this weekend. I keep in touch if i have some trouble in the C++ part. I'm also available to schedule some pair programming if needed.

fbraz3 commented:
@xezon I have updated the code on GitHub to change the release logic and add `VERSION_BUILDUSER` to the build toolchain. I also added the option to manually include build notes and known issues (as optional fields) for manual releases.

xezon commented on .github/workflows/build-toolchain.yml:
```diff
@@ -112,6 +116,10 @@ jobs:
           )
+          
+          if ("${{ inputs.release }}") {
+              $buildFlags += "-DVERSION_BUILDUSER='${{ inputs.release }}'"
```
What does `inputs.release` evaluate to? The variable name is strange.

xezon reviewed (commented):
Do you have a sample for what this workflow currently generates? Is it complete or is it missing anything?

OmniBlade commented on .github/workflows/weekly-release.yml:
```diff
@@ -0,0 +1,218 @@
+      - name: Create GitHub Release
+        uses: softprops/action-gh-release@v2
+        with:
+          tag_name: preview-${{ needs.get-date.outputs.date }}
```
Looking at this the format for the release name/tag should be preview-$DATE, but the most recent runs on your repo don't reflect that despite it being 2 weeks since you last pushed the branch? I'm guessing you need to push it to main for that to happen or something? Just wanted to check the output before I gace it the okay.

Also, do we want to call it "preview" which implies its just an early release of what could be final when in practice that won't be the case? Maybe just prefix "weekly" instead? Thoughts?

xezon replied:
"weekly" is also fine

Maybe "weekly-preview" :D

fbraz3 commented on .github/workflows/weekly-release.yml:
```diff
@@ -0,0 +1,218 @@
+        include:
+          - preset: "vc6"
+            tools: true
+            extras: true
```
Done

fbraz3 commented on .github/workflows/build-toolchain.yml:
```diff
@@ -110,6 +114,10 @@ jobs:
           )
+          
+          if ("${{ inputs.release-name }}") {
+              $buildFlags += "-DRELEASE_NAME='${{ inputs.release-name }}'"
```
done

fbraz3 commented:
@xezon , I've made some major changes to the codebase here.  

---

### Summary of Changes

The primary modification is how the version string is displayed on the options screen.

1.  **For standard CI builds** (from feature branches, etc.), the version string format remains unchanged:
    `Community Patch ~X by XXXX | exe:xxxxx ini:xxxx Version 1.04`

2.  **For release or pre-release builds** generated by the weekly build process (whether scheduled or manually triggered), the version string will now follow this pattern:
    `Community Patch by BUILD_USER release RELEASE_NAME - Version 1.04`

---

### How It Works

* The `weekly-release.yml` workflow sends `release-name`, `build_user`, and `build_loc` to the build toolchain.
* Standard CI processes do not send these parameters.
* The build toolchain treats these values as optional, defaulting to empty strings if they aren't provided.
* If the parameters are set, the toolchain passes them to CMake.
* The C++ code then reads these build parameters. If they are not blank, it changes the output string on the options screen accordingly.

<img width="802" height="627" alt="image" src="https://github.com/user-attachments/assets/44501149-d8e4-497f-ab8e-14e2ca7f0a94" />

OmniBlade reviewed (commented):
Maybe we drop the embedding of information from the CI for now as it is proving controvertial for exactly the best method to do so and just focus on producing weeklies as is from the CI? Embedding alternative build info from the CI can be a separate PR.

fbraz3 commented:
> Maybe we drop the embedding of information from the CI for now as it is proving controvertial for exactly the best method to do so and just focus on producing weeklies as is from the CI? Embedding alternative build info from the CI can be a separate PR.

I can do it, but in the I think the key point here is that the CI pipeline should be the single source of truth for versioning. This aligns with what’s considered best practice in the literature (Continuous Delivery – Humble & Farley, Build Automation – Fowler, and SemVer.org). Version numbers should reflect intentional release decisions, while commit hashes and build metadata are generated automatically. CI is the natural place to bring those together in a consistent way.

If we allow local builds to stamp arbitrary versions (date, hash, etc.), we risk drift: the same code might exist under multiple “versions” depending on who built it. By centralizing version stamping in CI, every official artifact can always be traced back to its exact commit and release tag, without mutating Git metadata. Local builds can still include fallback info (like “dev” or build time) for debugging, but those shouldn’t override what CI assigns.

fbraz3 commented:
> Maybe we drop the embedding of information from the CI for now as it is proving controvertial for exactly the best method to do so and just focus on producing weeklies as is from the CI? Embedding alternative build info from the CI can be a separate PR.

No problem to change the pr to only build in this first moment. will change the code and request a new review.

fbraz3 commented:
That’s true for any open source project — anyone can fork the repo and stamp whatever version string they want into their local builds. But that doesn’t reduce the value of CI-generated versioning. The real distinction is not whether the string can be replicated, but whether it represents an official artifact or just a local build.

The real value of CI versioning is giving us a single, consistent scheme for official artifacts. Local builds can still show commit/author info for debugging, but CI builds are what the community can trust as authoritative releases. That separation keeps things clear for both contributors and players.
